src/pretix/presale/checkoutflow.py

```diff
@@ -486,11 +488,34 @@ class AddOnsStep(CartMixin, AsyncAction, TemplateFlowStep):
     label = pgettext_lazy('checkoutflow', 'Add-on products')
     icon = 'puzzle-piece'
 
+    def _check_is_applicable(self, request):
+        self.request = request
+
+        # check whether addons are applicable
+        if get_cart(request).filter(item__addons__isnull=False).exists():
+            return True
+
+        # don't re-check whether cross-selling is applicable if we're already past the AddOnsStep
+        cur_step_identifier = request.resolver_match.kwargs.get('step')
+        is_past_this_step = any(step.identifier == cur_step_identifier for step in request._checkout_flow[request._checkout_flow.index(self) + 1:])
+        if is_past_this_step:
+            applicable = self.cart_session.get('_checkoutflow_addons_applicable', None)
+            if applicable is not None:
+                return applicable
+
+        # check whether cross-selling is applicable
+        applicable = self.cross_selling_is_applicable
+        self.cart_session['_checkoutflow_addons_applicable'] = applicable
```

Comment:
We were talking about the need of caching here. To think about this more, do you have first estimates on how many database queries this part takes?

Reply:
I managed to get a constant number of queries with regards to the number of items per category, categories in general, and number of cart positions, however for every cross-selling category that's actually rendered, a separate bunch of queries is performed by get_grouped_items. It might be possible to merge this somehow into a single call to get_grouped_items and split it up again afterwards, but I'm not sure whether that's worth the additional complexity... After all, I think in a usual shop there will be only one or a hand full of cross-selling categories visible in each order.